### ISSUE TITLE
Upgrade network so that valid block timestamps must be greater than the previous block

### ISSUE THREAD

**Issue by bitcartel** [zcash/zcash#1889]
Evaluate moving to an Ethereum style model of acceptable block timestamps, where in Ethereum:
1. A valid block timestamp must be greater than previous block timestamp
2. It must also be less than 900 seconds (15 minutes) into the future from current time.

http://ethereum.stackexchange.com/questions/5924/how-do-ethereum-mining-nodes-maintain-a-time-consistent-with-the-network

Ethereum nodes make use of NTP servers to stay in sync and adjust for clock drift:
http://www.pool.ntp.org/en/

**Comment by zookozcash** [comment]
I don't want to require Zcash nodes to rely on NTP, but I think this proposal would not require that of them. They could use some other mechanism of setting their clock, or they could actually use the timestamps from the Zcash blockchain to set their clock.

**Comment by str4d** [comment]
My initial objection to this was that out-by-hour issues are quite common. However, it should actually be relatively easy for miners and pools to do:

```
nTime = max(GetTime(), prev->nTime+1)
```

to ensure that their times are always locally valid. So concept ACK.

**Comment by str4d** [comment]
This could also be easily implemented from a consensus rule perspective, because the vast majority of miners are already submitting blocks that satisfy these rules. Implementation-wise it's going to take some work (as I know from I2P...)

**Comment by sammy007** [comment]
Smells like hard fork?

**Comment by wangchun** [comment]
I am disappointed to see a proposal like this. And it will NOT work and may open other issues. The easier solution might be to, instead of consider blocks [tip-17:tip-1], consider blocks [tip-18:tip-2].

**Comment by bitcartel** [comment]
@sammy007 @wangchun Just brainstorming at the moment.  All feedback welcome.

**Comment by bitcartel** [comment]
@vbuterin How has Ethereum's block timestamp worked out in production?  Any lessons to be learned?  If you could, what would you change today?

**Comment by str4d** [comment]
> Smells like hard fork?

My comment https://github.com/zcash/zcash/issues/1889#issuecomment-261824606 was saying that this *wouldn't* require a "hard" fork, because it would be making previously-valid blocks invalid. But even if it did, that's fine - we're already considering [other forking upgrades](https://z.cash/blog/zcash-evolution.html).

> There is only one requirement - rely on correct time.

Easier said than done, unfortunately... As discussed in the Chrome talk at RWC 2016, one of the top contributors to SSL certificate errors they see is discrepancies between client and server clocks, with IIRC a spike around an hour difference. So you either have to allow for significant discrepancies (as Bitcoin does), or run your own time-management system (as Ethereum and I2P do).

**Comment by zookozcash** [comment]
@wangchun: why won't the proposal in https://github.com/zcash/zcash/issues/1889#issue-190591769 work and what other issues could it raise?

**Comment by zookozcash** [comment]
> There is only one requirement - rely on correct time.

I strongly believe we should _not_ require this! I think it is both expensive/difficult/impossible to achieve, and also unnecessary.

**Comment by bitcartel** [comment]
To avoid DST issues, let's ensure:
- the code always asks the operating system for time in UTC
- the code always converts to and from unix timestamps in UTC rather than local time

We can use several sources of time when a new block arrives:
- network time, based on blockchain timestamps
- operating system clock, which typically syncs with NTP servers at boot time
- default NTP server to check against, like Ethereum with pool.ntp.org
- user-configuredNTP server
  - user has access to local, private or authenticated service
  - user has connected a hardware device as a source of time for `ntpd`
    - USB GPS dongle: https://blog.logentries.com/2015/07/adding-a-gps-time-source-to-ntpd
    - USB Atomic radio clock (MSF, DCF, WWVB radio time signals): https://www.meinbergglobal.com/english/products/usb-clocks.htm
    - GPS time server: http://www.css-timemachines.com/product/tm1000a-gps-time-server/

Some security concerns:
- NTP "Attacking the Network Time Protocol"
  - http://arstechnica.com/security/2015/10/new-attacks-on-network-time-protocol-can-defeat-https-and-create-chaos/
  - http://www.cs.bu.edu/~goldbe/papers/NTPattack.pdf
- Blockchain "TimeJacking"
  - http://culubas.blogspot.com/2011/05/timejacking-bitcoin_802.html
  - http://bitcoin.stackexchange.com/questions/3516/was-a-timejacking-attack-ever-performed

**Comment by zookozcash** [comment]
https://github.com/Electric-Coin-Company/infrastructure/issues/45

**Comment by vbuterin** [comment]
It's worked fairly well for us so far; setting timestamp to `min(now, parent.timestamp + 1)` is precisely what miners do. We do have a warning message that triggers when a node connects and it looks like it's out of sync; in my experience it's happened a few times but when it has, a lot of the time the timestamp wasn't just a little bit off, it was way off (eg. 8 hours) due to timezone changes. You probably don't want to rely on NTP exclusively; if a node is online you can rely on a combination of sources including NTP, timestamps of blocks, timestamps of peers on the network, etc.

This should not require a hard fork; the Satoshian scheme is strictly looser than the Ethereum scheme on both sides (if timestamps are increasing, then each timestamp is bigger than the previous median, and if a timestamp is less than the current time, it's less than the current time + 2 hours), so changing over would be a soft fork.

**Comment by daira** [comment]
This ticket was reported in http://www.newsbtc.com/2016/11/21/zcash-developers-contemplate-switch-ethereum-based-block-timestamping/ as if it were something we'd already decided on. Journalists and readers should be aware that we often discuss speculative changes on this issue tracker. It's an issue tracker, that's what it's for. :eyeroll:

**Comment by daira** [comment]
Tentative concept ACK for tightening the block timestamp rules.

**Comment by daira** [comment]
Note that Ethereum apparently doesn't have the 15-minute rule mentioned. Also, I think that we need to be clear on standard vs consensus rules. A rule based on local time can't be a consensus rule, it can only be a standard rule (which probably influenced Ethereum's decision to drop that rule in the Yellow Paper).

**Comment by bitcartel** [comment]
@daira The 15-minute rule made it into the white paper https://github.com/ethereum/wiki/wiki/White-Paper
> Check that the timestamp of the block is greater than that of the referenced previous block and less than 15 minutes into the future

But not into the code, here https://github.com/ethereum/go-ethereum/blob/master/core/block_validator.go#L220
> 	if header.Time.Cmp(parent.Time) != 1 {
		return BlockEqualTSErr
	}

and here https://github.com/ethereum/cpp-ethereum/blob/develop/libethcore/BlockHeader.cpp#L169
> if (m_timestamp <= _parent.m_timestamp)
     BOOST_THROW_EXCEPTION(InvalidTimestamp());

@vbuterin Is it documented anywhere why the 15 minute rule was dropped?  Thanks.

**Comment by nathan-at-least** [comment]
Upgrading to this rule requires care, and may need signaling for a soft-fork upgrade behavior, or some kind of hard-fork coordination, etc…

**Comment by tromer** [comment]
Note that keeping an adequate relation between block timestamps and the "real" (wall clock) time is important not just for difficulty adjustment, but also for transaction lock time to have useful semantics.

**Comment by zookozcash** [comment]
I'd like to have https://github.com/zcash/zcash/issues/1924 running for a while, reliable, and showing a history of past behavior before we deploy any change along these lines.

**Comment by zookozcash** [comment]
I think it is a mistake to conflate "my local clock is synced to 'the real time'" with "my local clock is running at approximately the right speed". I think the latter (my clock runs about approximately the right speed) is necessary and sufficient for the requirements and issues mentioned above (https://github.com/zcash/zcash/issues/1889#issuecomment-271427283, https://github.com/zcash/zcash/issues/1889#issuecomment-262072323, and "time-warping"). I think that the former assumption (my clock is approximately on the real time) is _impossible_ to ensure in theory (except with a blockchain!), and that relying on the former assumption will, in practice, give high-powered, large-scale attackers (i.e. those who can control widely-trusted NTP servers or attack the NTP protocol) the opportunity to attack our blockchain.

Just to hammer the point home, here it is again in compressed form:

property: "right time", necessary: "no", allows attacks: "yes"
property: "right speed", necessary: "yes", allows attacks: "no"

**Comment by tromer** [comment]
@zooko, transaction lock time may be expressed in two ways (depending on whether it's above the 500000000 threshold):
* as *block height*, for which it's plausible that "right speed" suffices (we just need difficulty adjustment, hence block pacing, to work well enough)
* as *UNIX timestamp*, for which it's necessary for the majority of miners to have an adequate "right time"

**Comment by bitcartel** [comment]
Some analysis of 42000 blocks:
- number of blocks where timestamp is earlier than previous block = 6271 (14.93%)
- number of affected blocks, where timestamp is earlier than 1+ blocks= 19654 (46.80%)
- longest chain of previous blocks, where a block's timestamp is earlier than = 5

**Comment by daira** [comment]
Recall that the proposed rules are:

1. A valid block timestamp must be greater than previous block timestamp
2. It must also be less than 900 seconds (15 minutes) into the future from current time.

Rule 2 cannot be a consensus rule because it mentions current time. So I'm going to interpret it as a rule about what blocks the *reference* miner will mine on top of.

A potential problem I see is that rule 2 is not incentive-compatible -- a miner gains no advantage from enforcing it, and in fact potentially loses the opportunity to build on a chain that had greater total work. So it is quite possible that non-reference miners will not enforce this rule.

This means that the block timestamps will tend to creep up relative to UTC, since there is nothing to prevent this from happening if only a minority of miners enforce rule 2.

**Comment by nathan-at-least** [comment]
Note, I've added #2022 which would make future empirical analysis of this kind of issue easier by providing useful local time data.

**Comment by jordanmack** [comment]
> 1. A valid block timestamp must be greater than previous block timestamp
> 2. It must also be less than 900 seconds (15 minutes) into the future from current time.

If rule 1 was a consensus rule, miners are forced to use `min(now, parent.timestamp + 1)` to avoid blocks being rejected. This will eventually inch the time closer to being back on track, but now the time lapsed between two blocks would again be inaccurate. If number 2 is is not a consensus rule, an incorrect date could lead to a scenario where the timestamp is thrown far into the future. You would then have a continuous string of blocks with a timestamp of `parent.timestamp + 1`, until the current date caught up... if ever. 

Enforcement of a tighter time window just complicates things. Rather than trying to enforce a small time window to prevent an incorrect local timestamp, what if you remove the consensus rules for the timestamp completely, then change the reference miner to use the network adjusted time for block timestamps and ignore the local time completely? To make the network adjusted time less prone to being thrown off by an incorrect date, only include timestamps within a standard deviation range to be included in the median. The network adjusted time should then be accurate within a few seconds of NTP. Even if a miner with a completely wrong date was to find a block, the block timestamp would still be correct. There would be no incentive for timejacking attempts since there would be no consensus rule in play to reject a bad timestamp.

**Comment by bitcartel** [comment]
Mainnet (posted above) https://github.com/zcash/zcash/issues/1889#issuecomment-271657370

Testnet first 42000 blocks:

- number of blocks mined where timestamp is earlier than one or more previous block = 279 (0.66%)
- total number of previous blocks = 294 (0.70%)
- longest chain of previous blocks = 5
- largest negative time delta between block and previous block = 202 secs

**Comment by zawy12** [comment]
Any upper limit you apply to timestamps should be reflected in a lower limit.  For example, you could follow the rule that the next timestamp is limited to +/- 750 seconds from the previous timestamp +150 seconds (+900 / -600).   If you don't allow the "negative" timestamp (-600 from previous timestamp) AND if miners can assign timestamps without a real-time limit from nodes, then a miner with > 20% of the network hashrate can drive the difficulty as low as he wants, letting everyone get blocks as fast as he wants, in less than a day.

A symmetrical limit on timestamps allows honest miner timestamps to completely erase the effect of bad timestamps. ( You do not need to wait 6 blocks for MTP like Zcash does in delaying the use of timestamps for difficulty, see footnote.  )  If you allow the symmetrical  "negative" timestamps, you do not need nodes to have the correct time with NTP or GPS unless miners collude with > 51% agreement on setting the timestamps further and further ahead of time to drive difficulty down.  It's a real possibility if miners decide they do not like a certain fork due to not providing them with enough fees.

But if your nodes have an accurate time, you do not need mining at all.  The only fundamental reason for mining is to act as a timestamp server to prevent double spending. 

BTC and ETH depend on nodes to limit the future time assigned to blocks. This seems like a bad joke. Zooko was the only one here who seemed to know there is something fishy about strong reliance on nodes having the correct time.  The extent to which BTC and ETH need those forward-time limits to be enforced by real time is the extent to which they do not need mining. 

Since gmaxwell (and apparently Satoshi) reject the idea of relying on state-sponsored and crash-able GPS,  NTP, or cellphone systems to eliminate the need for miners, the ideal solution is to have nodes that use a camera with a good zoom, known location, and accelerometer (if their camera is a cell phone not correctly mounted) to determine star position and to periodically calibrate their time based on that. @fluffypony was the only one I could get to "like" this idea on twitter.  Every honest desktop node could reject transactions with bad timestamps, within some small window like 1 minute (with good optics). Science like this does not need to ask for consensus.  Every node on his own could give the middle finger to every node that disagrees with him.  Nodes with correct time would naturally comprise the biggest network, all saying F-you to the miners.  Blocks could be 2 minutes apart and need only 1 confirmation.  Science began with looking at the stars and time is the only thing I can think of that computers can determine in isolation and then agree on without a trusted 3rd party. Colluding miners bullying us with > 51% hashrate into more total fees at the expense of security is a trusted 3rd party.

**Footnote:**
MTP does not stop a 25% attacker who can set timestamps > 4 blocks ahead if other miners are not allowed to assign a "negative" timestamp to eliminate the error in the next block.  But if you allow the "negatives" then MTP is not needed. Putting your tempering aside, this assumes you use 
```
next_D = avg(D's) * T / avg(solvetimes, allowing negative solvetime)
```
instead of 
```
next_D=sum(D's) * T / [max(Timestamps) - min(Timestamps) ]
```
because the N's of the denominator and number of the first equation do not cancel like you would think and hope (in order to use the second equation) when there are bad timestamps at the beginning and end of the window. With the MTP, your difficulty is delayed 5 blocks in responding to big ETH miners who jump on about twice a day.  That's like a gift to them at the expense of your constant miners.

Also, your tempered N=17 gives almost the same results as a straight average N=63.  I would use N=40 instead, without the tempering.  It should reduce the cheap blocks the big ETH miners are getting.

Your 16% / 32% limits are rarely reached due to the N=63 slowness.  This is good because it is a symmetry problem, although it would not be as bad as BCH.  Use "limit" and "1/limit" where  limit = X^(2/N) where N=63 for your current tempering and X = the size of the larger ETH attackers as a fraction of your total hashrate, which is about 3.  This allows the the fastest response up or down at N for a given X with 80% probability. Change the 2 to 3 to get a higher probability of an adequately-fast response.  The benefit is that it is a really loose  timestamp limit on individual values, as long as the aggregate is not too far from the expected range.

**Comment by zawy12** [comment]
Setting "negative" timestamps to previous timestamp is inserting an asymmetry that will open up an easy attack on coins not using MTP.   It may not cause a problem for Zcash who is using MTP, but it sets a bad example.  I sometimes have to battle with coins who are opening up an attack hole by inserting this asymmetry.  

The blue line below shows why using MTP is not needed if you don't erase the negative solvetimes.  Total time = maxTS - minTS will also not be as good as the blue line that uses total time = sum(solvetimes).  The MTP delay causes larger-than needed osciallations in Zcash because it gives away the first 5 blocks to attackers.

A 25% pool can lower any BTC clone's difficulty to 1/3 the correct difficulty if the algo erases negative timestamps, unless the difficulty starts at MTP.  This is NOT a reason to use MTP.  It shows not doing either mistake is best.

![negative_timestamps](https://user-images.githubusercontent.com/18004719/32165689-01e1882c-bd3a-11e7-9ce6-9a66f4ee6392.gif)

**Comment by daira** [comment]
(Changed the title because "upgrade" was making a positive value judgement.)

**Comment by daira** [comment]
It would definitely be a mistake for the future-dating threshold to be 15m, because a block interval greater than 15m can happen by chance, with significant probability over the expected lifetime of the protocol.

An exponential distribution with parameter λ has [cumulative distribution function](https://en.wikipedia.org/wiki/Exponential_distribution#Cumulative_distribution_function) 1 - e^{-λx}.
The block time approximately follows an exponential distribution with λ = 1/block_time if the latency of block propagation and verification is ignored (and the historical block time distribution is in good agreement with this).

So, the probability of a span greater than x, is about e^{-x / block_time}.

Here is the threshold time in minutes, vs the expected time in years before that threshold happens by chance (post-Blossom):

10m 0.007y
15m 0.387y
20m 21.119y
25m 1153.047y
30m 62954.222y
35m 3437184.064y

as produced by this code:
```
from math import e
for x in xrange(10, 40, 5):
    print "%2dm %0.3fy" % (x, e**(x*60/75.0)*75.0/3600/24/365.25)
```

In other words we probably shouldn't reduce the threshold below 30m.

**Comment by zawy12** [comment]
**Summary:**
I assume we're talking about reducing the **[7200 second limit](https://github.com/zcash/zcash/blob/0a31c163c68b9875c101df544c7c167952ae4f43/src/main.cpp#L3696)** to 15 minutes but @daira is interpreting it differently.  If my assumption is correct, rules 1 and 2 seem fine as long as the **[revert rule](https://github.com/zcash/zcash/blob/8dafb9c6cee260f095a85bb693726600329b9c98/src/timedata.cpp#L82)** is reduced from 70 to 8 minutes.

**FTL and revert rule:**
BTC's Future Time Limit (FTL) of 7200 seconds is the "allowed error" in timestamps after current time. "Current" time is the median of peer time **or** the node's time if the median differs from node's time by > 4200 seconds (the 70 minute "revert rule").  The median is a consensus mechanism that's subject to 33% Sybil attacks on each miner's current time, but potentially overridden by an oracle (his node time). This stops Sybil attacks that could lower difficulty by artificially increasing the FTL to something a lot larger than 7200+4200). 

**Correct revert rule or remove median of peer time**
I do not see a reason to use median of peer time. A Sybil attack on it could split the network hashrate, potentially greatly reducing the hashrate needed to get blocks (including a selfish mine with double spends), at least if it were not for the problem and benefits of pool centralization.  If it is used I would  keep the revert rule like BTC where it is about 1/2 of the FTL, i.e. 8 minutes if FTL=15 minutes.

A decentralization purist would not use NTP or even a GPS method to get time, but would advise miners to get templates from nodes that independently choose how they get world time. If miners disagree on time beyond the FTL, network hashrate is split, but an attack would have to have >50% hashrate (aka >25%) to win, which is not the case in the Sybil attack above which could split hashrate 4 ways and thereby need only 6.25% of hashrate to win a chain work race.

**Consensus always requires a 2nd consensus mechanism or an oracle to get time**
All distributed consensus mechanisms require an oracle for time (or time rate limits) to dictate when consensus rounds begin and/or end. For example, a simple mechanism might require >50% of the known participants to respond and agree in a yes/no round before declaring a consensus (as opposed to declaring a partition and restarting). Nakamoto consensus turns it upside by using the block solve time to estimate the number of participants, calling the sequence of rounds with the largest participation (chain work) the winner, resolving partitions. This is the only known truly permissionless consensus mechanism (there is no "voter" registration, declaration, or identification before a round).  

**Sequential timestamps**
If timestamps are required to be sequential, it removes the need to use median of past 11 timestamps as the most recent timestamp in difficulty calculation which will reduce a small but persistent oscillation and allow it to respond faster without causing other problems.  If the MTP(11) were not used for anything else then the sequential +1 rule and difficulty calculation could both be implemented by just changing the 11 to 1.
